opal/mca/threads/qthreads/threads_qthreads_mutex.h

```diff
@@ -42,187 +42,144 @@
-            .m_recursive = 1, .m_lock_debug = 0, .m_lock_file = NULL, .m_lock_line = 0,   \
-            .m_lock_atomic = OPAL_ATOMIC_LOCK_INIT,                                       \
-        }
-#else
-#    define OPAL_RECURSIVE_MUTEX_STATIC_INIT                                              \
-        {                                                                                 \
-            .super = OPAL_OBJ_STATIC_INIT(opal_mutex_t), .m_lock = OPAL_ATOMIC_LOCK_INIT, \
-            .m_recursive = 1, .m_lock_atomic = OPAL_ATOMIC_LOCK_INIT,                     \
-        }
-#endif
-
-/************************************************************************
- *
- * mutex operations (non-atomic versions)
- *
- ************************************************************************/
+#define OPAL_THREAD_INTERNAL_MUTEX_INITIALIZER           OPAL_ATOMIC_LOCK_INIT
+#define OPAL_THREAD_INTERNAL_RECURSIVE_MUTEX_INITIALIZER OPAL_ATOMIC_LOCK_INIT
 
-static inline void opal_mutex_create(struct opal_mutex_t *m)
+static inline int opal_thread_internal_mutex_init(opal_thread_internal_mutex_t *p_mutex,
+                                                  bool recursive)
 {
-    opal_threads_ensure_init_qthreads();
-
-    opal_atomic_lock_init(&m->m_lock, 0);
-    opal_atomic_lock_init(&m->m_lock_atomic, 0);
-    m->m_recursive = 0;
-#if OPAL_ENABLE_DEBUG
-    m->m_lock_debug = 0;
-    m->m_lock_file = NULL;
-    m->m_lock_line = 0;
-#endif
+    opal_atomic_lock_init(p_mutex, 0);
+    return OPAL_SUCCESS;
 }
 
-static inline void opal_mutex_recursive_create(struct opal_mutex_t *m)
+static inline void opal_thread_internal_mutex_lock(opal_thread_internal_mutex_t *p_mutex)
 {
     opal_threads_ensure_init_qthreads();
 
-    opal_atomic_lock_init(&m->m_lock, 0);
-    opal_atomic_lock_init(&m->m_lock_atomic, 0);
-    m->m_recursive = 1;
-#if OPAL_ENABLE_DEBUG
-    m->m_lock_debug = 0;
-    m->m_lock_file = NULL;
-    m->m_lock_line = 0;
-#endif
+    int ret = opal_atomic_trylock(p_mutex);
+    while (0 != ret) {
+        qthread_yield();
+        ret = opal_atomic_trylock(p_mutex);
+    }
 }
 
-static inline int opal_mutex_trylock(opal_mutex_t *m)
+static inline int opal_thread_internal_mutex_trylock(opal_thread_internal_mutex_t *p_mutex)
 {
     opal_threads_ensure_init_qthreads();
 
-    int ret = opal_atomic_trylock(&m->m_lock);
+    int ret = opal_atomic_trylock(p_mutex);
     if (0 != ret) {
         /* Yield to avoid a deadlock. */
         qthread_yield();
     }
     return ret;
 }
 
-static inline void opal_mutex_lock(opal_mutex_t *m)
+static inline void opal_thread_internal_mutex_unlock(opal_thread_internal_mutex_t *p_mutex)
 {
     opal_threads_ensure_init_qthreads();
 
-    int ret = opal_atomic_trylock(&m->m_lock);
-    while (0 != ret) {
-        qthread_yield();
-        ret = opal_atomic_trylock(&m->m_lock);
-    }
+    opal_atomic_unlock(p_mutex);
+    /* For fairness of locking. */
+    qthread_yield();
```

Comment:
Thank you for reviewing this PR! It is to avoid a deadlock. It is a typical pattern that does something like this:
```c
while (p_request->completed) {
  mutex_lock(&lock);
  progress_something();
  mutex_unlock(&lock);
}
```
If it is the case, sometimes the thread needs to yield "after" unlocking this lock to avoid a deadlock (i.e., yield *inside* this `progress_something()` might not be sufficient if another thread that this thread depends on needs to take this lock).

Of course, this yield would be inefficient in some/many cases, but prior to understanding everything, I would like to keep this yield.